### PR TITLE
Add Indicator data director

### PR DESCRIPTION
**Pull request by KhadijaMahanga**
## Checklist

- [ ]  🚀 is the code ready to be merged and go live?
- [ ]  🛠 does it work (build) locally

### Pull Request

- [ ]  📰 good title
- [ ]  📝good description
- [ ]  🔖 issue linked
- [ ]  📖 changelog filled out

### Commits

- [ ]  commits are clean
- [ ]  commit messages are clean

### Code Quality

- [ ]  🚧 no commented out code
- [ ]  🖨 no unnecessary logging
- [ ]  🎱 no magic numbers
- [ ]  black was run locally (as part of the pre-commit hook)

### Testing

- [ ]  ✅ added (appropriate) unit tests
- [ ]  💢 edge cases in tests were considered
- [ ]  ✅ ran tests locally & are passing

### CODE REVIEW

File: wazimap_ng/datasets/admin/forms/__init__.py [hunk 1]
```diff
@@ -1,2 +1,3 @@
 from .dataset_admin_form import DatasetAdminForm
-from .indicator_admin_form import IndicatorAdminForm
\ No newline at end of file
+from .indicator_admin_form import IndicatorAdminForm
+from .indicator_drector_admin_form import IndicatorDirectorForm
```
Comment (kilemensi) [inline wazimap_ng/datasets/admin/forms/__init__.py:3]:
```suggestion
from .indicator_drector_admin_form import IndicatorDirectorForm

```

File: wazimap_ng/datasets/admin/forms/indicator_drector_admin_form.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from django import forms
+from ... import models
+
+class IndicatorDirectorForm(forms.Form):
+    datasetfile = forms.ModelChoiceField(queryset=models.DatasetFile.objects.all(), required=True)
```
Comment (kilemensi) [inline wazimap_ng/datasets/admin/forms/indicator_drector_admin_form.py:5]:
```suggestion
    dataset_file = forms.ModelChoiceField(queryset=models.DatasetFile.objects.all(), required=True)
```
So a single dataset file can be used in multiple directors?

File: wazimap_ng/general/static/js/indicatordata-admin.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        function selectOnChange(val) {
+            if (val == "1") {
+                showNode("indicatordata_form", "none");
+                var link = '/admin/datasets/indicatordata/upload';
+                window.location.href = link;
+            } else if (val == "2") {
+                showNode("indicatordata_form", "block");
+                //showNode("add_method_div", "none");
```
Comment (kilemensi) [inline wazimap_ng/general/static/js/indicatordata-admin.js:20]:
```suggestion
```

File: wazimap_ng/general/templates/admin/indicatordata_change_list.html [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        <select id="add_method">
+            <option value="">----------------</option>
+            <option value="1">Upload Indicator Director</option>
+            <option value="2">Create from Dataset </option>
+        </select>
+    </div>
+
+    </div>
+    {{ block.super }}
+{% endblock %}
```
Comment (kilemensi) [inline wazimap_ng/general/templates/admin/indicatordata_change_list.html:16]:
```suggestion
{% endblock %}

```

File: wazimap_ng/general/templates/admin/indicatordata_upload_director.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+{% load i18n admin_urls %}
+{% block content %}
+  <form enctype="multipart/form-data" action="" method="post">
+    {% csrf_token %}
+    <table>
+    {{ form.as_table }}
+    </table>
+    <input type="submit" value="Submit">
+  </form>
+{% endblock %}
```
Comment (kilemensi) [inline wazimap_ng/general/templates/admin/indicatordata_upload_director.html:11]:
```suggestion
{% endblock %}

```

File: wazimap_ng/general/static/js/indicatordata-admin.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+(function($) {
+    jQuery(document).ready(function($) {
+        //check if add_method exist
+        var el = document.getElementById("add_method_div");
+        if (el) {
+            var sel = document.getElementById("add_method");
+            sel.onchange = function() { selectOnChange(this.value)};
```
Comment (kilemensi) [inline wazimap_ng/general/static/js/indicatordata-admin.js:7]: Is it possible to pass the element itself as as value? That way we wouldn’t need the how `showNode` method

Alternatively, can we get the source/target from event itself?

**Review by kilemensi — COMMENTED** [review]
👍🏼

File: wazimap_ng/datasets/admin/indicator_data_admin.py [hunk 7]
```diff
@@ -38,8 +49,80 @@ def parent(self, obj):
...
+    def upload_indicator_director(self, request):
+
+        if request.method == 'POST':
+            form = IndicatorDirectorForm(request.POST, request.FILES)
+            if form.is_valid():
+                indicator_director = request.FILES["indicator_director"]
+                dataset = form.cleaned_data["dataset"]
+                logger.debug(f" Uploaded Indicator director file: {indicator_director}")
+                
+                indicator_indicator_json = indicator_director.read()
```
Comment (esirK) [inline wazimap_ng/datasets/admin/indicator_data_admin.py:79]: This works but from Django [docs](https://docs.djangoproject.com/en/3.1/ref/files/uploads/#django.core.files.uploadedfile.UploadedFile.read), if the file is too large, we might get some issues.

File: wazimap_ng/datasets/admin/indicator_data_admin.py [hunk 8]
```diff
@@ -1,16 +1,27 @@
...
+from django.urls import path
+from django.http import HttpResponseRedirect
 
 from django_json_widget.widgets import JSONEditorWidget
+from django.template.response import TemplateResponse
+
+from django_q.tasks import async_task
 
 from .base_admin_model import DatasetBaseAdminModel
+from .. import hooks
```
Comment (esirK) [inline wazimap_ng/datasets/admin/indicator_data_admin.py:14]: we can have this like `from .. import models, hooks` right?

File: wazimap_ng/datasets/admin/forms/indicator_director_admin_form.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+from django import forms
+from ... import models
+
+class IndicatorDirectorForm(forms.Form):
+    dataset = forms.ModelChoiceField(queryset=models.Dataset.objects.all(), required=True)
+    indicator_director = forms.FileField(required=True)
+
+
```
Comment (kilemensi) [inline wazimap_ng/datasets/admin/forms/indicator_director_admin_form.py:8]:
```suggestion
```

File: wazimap_ng/datasets/admin/indicator_data_admin.py (hunk 7 above)
Comment (kilemensi) [inline wazimap_ng/datasets/admin/indicator_data_admin.py:79]:
```suggestion
                indicator_director_json = indicator_director.read()
```
Or I'm missing something?

File: wazimap_ng/datasets/tasks/process_indicator_data_director.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+@transaction.atomic
+def process_indicator_data_director(indicator_director, dataset, **kwargs):
+    logger.debug(f"process uploaded director file: {dataset}")
+
+    sorter = Sorter()
+    data = json.loads(indicator_director)
+
+    indicator_name = list(data)[0]
+    primary_groups = data[indicator_name]["subindicators"]
```
Comment (kilemensi) [inline wazimap_ng/datasets/tasks/process_indicator_data_director.py:27]: Is this equivalent to:
```python

     for indicator_name, indicator in data.items():
        primary_groups = indicator["subindicators"]
```
?

File: wazimap_ng/datasets/tasks/process_indicator_data_director.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+        models.IndicatorData.objects.bulk_create(datarows, 1000)
+
+        return {
+            "model": "indicator",
+            "name": indicator.name,
+            "id": indicator.id,
+        }   
+    else:
+        raise Exception("subindicators not in dataset groups")
```
Comment (kilemensi) [inline wazimap_ng/datasets/tasks/process_indicator_data_director.py:67]: New line

File: wazimap_ng/profile/serializers/indicator_data_serializer.py [hunk 12]
```diff
@@ -164,7 +164,7 @@ def IndicatorDataSerializer(profile, geography):
         lambda x: x["jsdata"]["subindicators"]
     )
 
-    d_children = pivot(d_children, [0, 1, 2, 3, 4, 5, 8, 7, 6])
+    d_children = pivot(d_children, [0, 1, 2, 3, 4, 5, 8, 9, 7, 6])
```
Comment (kilemensi) [inline wazimap_ng/profile/serializers/indicator_data_serializer.py:167]: What does this magic mean?

**Review by kilemensi — APPROVED** [review]
🚀 @KhadijaMahanga . I think it works well and we can use it to start discussions upstream. Lets double check the use of **indicator director** especially in **indicator data director**, it can get confusing. I _think_ it should be **_indicator director data_**

What I would have liked to see or see explained:

1. Format of Indicator Director file,
2. Can Indicator Director be modified?
3. Can we have any test for Indicator Director.

File: wazimap_ng/datasets/tasks/process_indicator_data_director.py (hunk 10 above)
Comment (KhadijaMahanga) [reply wazimap_ng/datasets/tasks/process_indicator_data_director.py:27]: @kilemensi  right! `.... = data.items()[0]` ...but with python2. Python3 return a generator and can't be indexed

File: wazimap_ng/profile/serializers/indicator_data_serializer.py (hunk 12 above)
Comment (KhadijaMahanga) [reply wazimap_ng/profile/serializers/indicator_data_serializer.py:167]: This magic re-arranges position of indicator data keys. The pivot function is explained here https://github.com/CodeForAfrica/wazimap-ng/blob/f334b8ebfa4ed94746245fd49ea49ceb18951655/wazimap_ng/utils.py#L514

**Review by esirK — APPROVED** [review]
👍

File: wazimap_ng/datasets/admin/forms/indicator_director_admin_form.py [hunk 13]
```diff
@@ -0,0 +1,6 @@
+from django import forms
+from ... import models
+
+class IndicatorDirectorForm(forms.Form):
+    dataset_file = forms.FileField(required=True)
+    indicator_director = forms.FileField(required=True)
```
Comment (kilemensi) [inline wazimap_ng/datasets/admin/forms/indicator_director_admin_form.py:6]:
```suggestion
    indicator_director_file = forms.FileField(required=True)
```

File: wazimap_ng/datasets/admin/indicator_data_admin.py [hunk 14]
```diff
@@ -38,8 +50,84 @@ def parent(self, obj):
...
+
+    def upload_indicator_director(self, request):
+        if not request.user.is_authenticated:
+            return redirect('/admin/login?next=%s' % (request.path))
+
+        if request.method == 'POST':
+            form = IndicatorDirectorForm(request.POST, request.FILES)
+            if form.is_valid():
+                indicator_director = form.cleaned_data.get("indicator_director", None)
+                dataset_file = form.cleaned_data.get("dataset_file", None)
```
Comment (kilemensi) [inline wazimap_ng/datasets/admin/indicator_data_admin.py:79]:
```suggestion
                form_data = form.cleaned_data
                indicator_director_file = form_data.get("indicator_director_file", None)
                dataset_file = form_data.get("dataset_file", None)
```

File: wazimap_ng/datasets/admin/indicator_data_admin.py [hunk 15]
```diff
@@ -38,8 +50,84 @@ def parent(self, obj):
...
+
+                if dataset_file and indicator_director:
+                    datasetfile_obj = models.DatasetFile.objects.create(
```
Comment (kilemensi) [inline wazimap_ng/datasets/admin/indicator_data_admin.py:82]:
```suggestion
                    dataset_file_obj = models.DatasetFile.objects.create(
```

File: wazimap_ng/datasets/admin/indicator_data_admin.py [hunk 16]
```diff
@@ -38,8 +50,84 @@ def parent(self, obj):
...
+                        name=dataset_file.name,
+                        document=dataset_file
+                    )
+                    indicator_director_json = indicator_director.read()
```
Comment (kilemensi) [inline wazimap_ng/datasets/admin/indicator_data_admin.py:86]:
```suggestion
                    indicator_director_json = json.load(indicator_director_file)
```

File: wazimap_ng/datasets/tasks/process_indicator_data_director.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+def process_indicator_data_director(data, datasetfile_obj, **kwargs):
+    logger.debug(f"process uploaded director file: {dataset}")
+
+    sorter = Sorter()
+    
+    indicator_name = list(data)[0]
+    data_obj = data[indicator_name]
+    primary_groups = data_obj["subindicators"]
+
+    dataset = process_dataset_file(data_obj, datasetfile_obj, indicator_name)
```
Comment (kilemensi) [inline wazimap_ng/datasets/tasks/process_indicator_data_director.py:48]: Looks like `process_dataset_file` loads `Profile` and `GeographyHierarchy` objects from DB. I don't see where we're passing the ids for these (or should the dataset file define these?).

**Review by kilemensi — COMMENTED** [review]
👍🏾 ... I'm stuck at the process of dataset file: How do we select / specify `Profile` and `GeographyHierarchy` for the dataset to use?

File: wazimap_ng/datasets/tasks/process_indicator_data_director.py (hunk 17 above)
Comment (KhadijaMahanga) [reply wazimap_ng/datasets/tasks/process_indicator_data_director.py]: @kilemensi the indicator_dirctor json should define these.
```
{
      indicator_name: {
              "subindicators": ["a"],
              "profile": 1,
              "geography_hierarchy": 1
     }
}
```

Comment (kilemensi) [reply wazimap_ng/datasets/tasks/process_indicator_data_director.py]: Why though? Can't they be selected from a drop down or something when submitting the director file?

Comment (KhadijaMahanga) [reply wazimap_ng/datasets/tasks/process_indicator_data_director.py]: @kilemensi that's possible too.

Comment (KhadijaMahanga) [reply wazimap_ng/datasets/tasks/process_indicator_data_director.py]: @kilemensi  but we're literally repeating the dataset admin page. Why don't we just select a dataset from a dataset as before

Comment (kilemensi) [reply wazimap_ng/datasets/tasks/process_indicator_data_director.py]: I'm not sure I follow @KhadijaMahanga ...

**Review by kilemensi — COMMENTED** [review]
LGTM @KhadijaMahanga 

--------------------------

1. Can we arrange 👇🏾 form: 1. Profile, 2. Geography, 3. Dataset, and 4. Indicator Data Director
Screenshot from 2021-03-17 09-36-36
2. There was a requirement or suggestion that the created dataset (or indicator, can't remember) should not be reusable afterwards. How/where are we handling that.

**Comment by KhadijaMahanga** [comment]
@kilemensi  On this `There was a requirement or suggestion that the created dataset (or indicator, can't remember) should not be reusable afterwards. How/where are we handling that.` I have added  "restricted" among permission_type and using that to exclude from queryset. Is it a good approach?

**Comment by kilemensi** [comment]
> @kilemensi On this  ... Is it a good approach?

I'm not sure... I think it works and we can ping OpenUp for review but I'm also thinking this is different from `permission` i.e. visibility. It's more about if that Dataset is open or closed for reuse / modifications. I think a Dataset can still be public i.e. fully visible but closed i.e. should not be used to create new indicators i.e. we should add a field to Dataset to keep track of this (separately from / in addition to `permission`).

**Comment by kilemensi** [comment]
> > @kilemensi On this  ... Is it a good approach?
> 
> I'm not sure... I think it works and we can ping OpenUp for review but I'm also thinking this is different from `permission` i.e. visibility. It's more about if that Dataset is open or closed for reuse / modifications. I think a Dataset can still be public i.e. fully visible but closed i.e. should not be used to create new indicators i.e. we should add a field to Dataset to keep track of this (separately from / in addition to `permission`).

What was the final decision on this @KhadijaMahanga ?
